src/main/java/hudson/plugins/ec2/EC2CleanupOrphanedNodes.java

```diff
@@ -0,0 +1,276 @@
+                updateLocalInstancesTag(connection, localConnectedEC2Instances, remoteInstancesIds, cloud);
+
+        remoteInstances.stream()
+                // exclude instances that just got updated
+                .filter(remote -> !updatedInstances.contains(remote.instanceId()))
+                .filter(this::isOrphaned)
+                .forEach(remote -> terminateInstance(remote, connection));
+    }
+
+    private List<EC2Cloud> getClouds() {
+        return Jenkins.get().clouds.getAll(EC2Cloud.class);
+    }
+
+    /**
+     * Returns a list of all EC2 instances in states (running, pending, or stopping) AND with the tags
+     * jenkins_server_url and jenkins_slave_type
+     * These are all the instances that are created by the EC2 plugin
+     */
+    private Set<Instance> getAllRemoteInstanceIds(Ec2Client connection, EC2Cloud cloud) throws SdkException {
+        Set<Instance> instanceIds = new HashSet<>();
+
+        String nextToken = null;
+
+        do {
+            DescribeInstancesRequest.Builder requestBuilder = DescribeInstancesRequest.builder()
+                    .maxResults(500)
+                    .filters(
+                            Filter.builder()
+                                    .name("instance-state-name")
+                                    .values(
+                                            InstanceState.RUNNING.getCode(),
+                                            InstanceState.PENDING.getCode(),
+                                            InstanceState.STOPPING.getCode())
+                                    .build(),
+                            tagFilter(EC2Tag.TAG_NAME_JENKINS_SLAVE_TYPE),
+                            tagFilter(EC2Tag.TAG_NAME_JENKINS_SERVER_URL));
+
+            requestBuilder.nextToken(nextToken);
+
+            DescribeInstancesResponse result = connection.describeInstances(requestBuilder.build());
+
+            for (Reservation r : result.reservations()) {
+                instanceIds.addAll(new HashSet<>(r.instances()));
+            }
+
+            nextToken = result.nextToken();
+        } while (nextToken != null);
+
+        LOGGER.fine(() -> "Found " + instanceIds.size() + " remote instance ID(s) for cloud: " + cloud.getDisplayName()
+                + ". Instance IDs: "
+                + instanceIds.stream().map(Instance::instanceId).collect(Collectors.joining(", ")));
+        return instanceIds;
+    }
+
+    /**
+     * Returns a list of EC2 agent instance IDs connected to Jenkins.
+     */
+    private Set<String> getConnectedAgentInstanceIds(EC2Cloud cloud) {
+        return Jenkins.get().getNodes().stream()
+                .filter(EC2AbstractSlave.class::isInstance)
+                .map(EC2AbstractSlave.class::cast)
+                .filter(node -> cloud.equals(node.getCloud()))
+                .peek(node -> LOGGER.fine(
+                        () -> "Connected agent: " + node.getNodeName() + ", Instance ID: " + node.getInstanceId()))
```

Comment:
Usage of this method can be problematic. Here it looks fine as all elements are collected, but it depends on the terminal operation.